api-server/web-server/src/api/v1.rs

```diff
@@ -315,8 +317,77 @@ async fn get_transaction(
         ))
 }
 
+pub async fn feerate<T: ApiServerStorage>(
+    Query(params): Query<BTreeMap<String, String>>,
+    State(state): State<ApiServerWebServerState<Arc<T>, Arc<impl TxSubmitClient>>>,
+) -> Result<impl IntoResponse, ApiServerWebServerError> {
+    const REFRESH_INTERVAL_SEC: u64 = 30;
+    const IN_TOP_X_MB: &str = "in_top_x_mb";
+    const DEFAULT_IN_TOP_X_MB: usize = 5;
+    let in_top_x_mb = params
+        .get(IN_TOP_X_MB)
+        .map(|str| usize::from_str(str))
+        .transpose()
+        .map_err(|_| ApiServerWebServerClientError::InvalidInTopX)?
+        .unwrap_or(DEFAULT_IN_TOP_X_MB);
+
+    let feerate_points = &state.cached_values.feerate_points;
+
+    let feerate_points: BTreeMap<_, _> = {
+        let current_time = get_time();
+        if (feerate_points.read().expect("should not fail normally").0
+            + Duration::from_secs(REFRESH_INTERVAL_SEC))
+        .expect("no overflow")
+            < current_time
+        {
+            let new_feerate_points = {
+                state.rpc.get_feerate_points().await.map_err(|e| {
+                    logging::log::error!("internal error: {e}");
+                    ApiServerWebServerError::ServerError(
+                        ApiServerWebServerServerError::InternalServerError,
+                    )
+                })?
+            };
+
+            let mut guard = feerate_points.write().expect("should not fail normally");
+            guard.0 = current_time;
+            guard.1 = new_feerate_points;
+            guard
+                .1
+                .iter()
+                .map(|(size, feerate)| (*size, Amount::from_atoms(feerate.atoms_per_kb())))
+                .collect()
+        } else {
+            feerate_points
+                .read()
+                .expect("should not fail normally")
+                .1
+                .iter()
+                .map(|(size, feerate)| (*size, Amount::from_atoms(feerate.atoms_per_kb())))
+                .collect()
+        }
+    };
+
+    let (min_size, max_feerate) = feerate_points.first_key_value().expect("not empty");
+    let (max_size, min_feerate) = feerate_points.last_key_value().expect("not empty");
+    let feerate = match in_top_x_mb {
+        _ if in_top_x_mb <= *min_size => *max_feerate,
+        _ if in_top_x_mb >= *max_size => *min_feerate,
+        _ => mempool::find_interpolated_value(&feerate_points, in_top_x_mb).ok_or_else(|| {
```

Comment:
Looks `in_top_x_mb` is matched but not inspected in patterns. Seems to me `if..else` cascade would be more idiomatic here. If you still want to use match for some reason (it kind of is more compact with less syntactic noise), just match over the unit `()`.